File: ios/bridge/ExpoBridgeHistory.swift
import Foundation

@available(iOS 16.0, *)
enum ExpoBridgeHistory {
  static func getRecentlyPlayedResources(service: HistoryService) async throws -> [String: Any] {
    try await AppleMusicBridgeError.rethrow {
      let tracks = try await service.getRecentlyPlayedResources()
      return BridgeResponses.recentlyPlayedResources(tracks)
    }
  }

  static func getRecentlyPlayedTracks(service: HistoryService, options: NSDictionary) async throws -> [String: Any] {
    try await AppleMusicBridgeError.rethrow {
      let pagination = BridgePagination(from: options)
      let songs = try await service.getRecentlyPlayedTracks(options: pagination)
      return BridgeResponses.songs(songs)
    }
  }

  static func getHeavyRotation(service: HistoryService, options: NSDictionary) async throws -> [String: Any] {
    try await AppleMusicBridgeError.rethrow {
      let pagination = BridgePagination(from: options)
      let items = try await service.getHeavyRotation(limit: pagination.limit)
      return BridgeResponses.recentItems(items)
    }
  }

  static func getRecentlyPlayedStations(service: HistoryService, options: NSDictionary) async throws -> [String: Any] {
    try await AppleMusicBridgeError.rethrow {
      let pagination = BridgePagination(from: options)
      let stations = try await service.getRecentlyPlayedStations(limit: pagination.limit)
      return BridgeResponses.stations(stations)
    }
  }

  static func getRecentlyAdded(service: HistoryService, options: NSDictionary) async throws -> [String: Any] {
    try await AppleMusicBridgeError.rethrow {
      let pagination = BridgePagination(from: options)
      let items = try await service.getRecentlyAdded(limit: pagination.limit, offset: pagination.offset)
      return BridgeResponses.recentItems(items)
    }
  }
}

File: android/src/main/java/expo/modules/applemusic/bridge/ExpoBridgeHistory.kt
package expo.modules.applemusic.bridge

import expo.modules.applemusic.AndroidHistoryService
import expo.modules.applemusic.BridgeResponses
import expo.modules.applemusic.PaginationOptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.ModuleDefinitionBuilder

internal fun ModuleDefinitionBuilder.registerHistoryBridge(
  historyService: () -> AndroidHistoryService,
) {
  AsyncFunction("getRecentlyPlayedResources") Coroutine { musicUserToken: String ->
    BridgeResponses.recentlyPlayedResources(historyService().getRecentlyPlayedResources(musicUserToken))
  }

  AsyncFunction("getRecentlyPlayedTracks") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.songs(historyService().getRecentlyPlayedTracks(musicUserToken, pagination))
  }

  AsyncFunction("getHeavyRotation") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.recentItems(historyService().getHeavyRotation(musicUserToken, pagination))
  }

  AsyncFunction("getRecentlyPlayedStations") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.stations(historyService().getRecentlyPlayedStations(musicUserToken, pagination))
  }

  AsyncFunction("getRecentlyAdded") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.recentItems(historyService().getRecentlyAdded(musicUserToken, pagination))
  }
}
